core/src/main/scala/com/mesosphere/usi/core/scalaapi/SchedulerAdapter.scala

```diff
@@ -0,0 +1,170 @@
+package com.mesosphere.usi.core.scalaapi
+
+import akka.stream.scaladsl.{Flow, Sink, SinkQueueWithCancel, Source, SourceQueueWithComplete}
+import akka.stream._
+import akka.{Done, NotUsed}
+import com.mesosphere.usi.core.Scheduler.{SpecInput, StateOutput}
+import com.mesosphere.usi.core.models._
+
+import scala.concurrent.{ExecutionContext, Future, Promise}
+import scala.util.{Failure, Success}
+
+/**
+  * SchedulerAdapter provides a set of simple interfaces to the scheduler flow.
+  * @param schedulerFlow
+  * @param mat
+  */
+class SchedulerAdapter(schedulerFlow: Flow[SpecInput, StateOutput, NotUsed])(
+    implicit mat: Materializer,
+    ec: ExecutionContext) {
+
+  /**
+    * Represents the scheduler as a Sink and Source.
+    *
+    * This method will materialize the scheduler first, then Sink and Source can be materialized independently, but only once.
+    *
+    * @param specsSnapshot Snapshot of the current specs
+    * @return Snapshot of the current state, as well as Source which produces StateEvents and Sink which accepts SpecEvents
+    */
+  def asSourceAndSink(specsSnapshot: SpecsSnapshot = SpecsSnapshot.empty)
+    : (Future[StateSnapshot], Source[StateEvent, NotUsed], Sink[SpecUpdated, NotUsed]) = {
+
+    val (stateQueue, stateSource) = Source.queue[StateEvent](1, OverflowStrategy.backpressure).preMaterialize()
+
+    val (specQueue, specSink) = Sink.queue[SpecUpdated]().preMaterialize()
+
+    val stateSnapshotPromise = Promise[StateSnapshot]()
+
+    val killSwitch = KillSwitches.shared("SchedulerAdapter.asSourceAndSink")
+
+    // We need to handle the case when source is cancelled or failed
+    stateQueue.watchCompletion().onComplete {
+      case Success(_) =>
+        killSwitch.shutdown()
+      case Failure(cause) =>
+        killSwitch.abort(cause)
+    }
+
+    Source.maybe.prepend {
```

Comment:
So, basically, this is a `Source.fromQueue` method, right? 😊 Should we extract it and call it what it is. I mean, since we have to use this workaround.

Reply:
If we extract it, this code might pop up in other places :) I can make it private though.